i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,248 @@
+            pos_seq_k = torch.arange(time_dim_size, device=input_tensor.device)
+
+            distance_mat = pos_seq_k[None, :] - pos_seq_q[:, None]
+            distance_mat_clipped = torch.clamp(distance_mat, -self.rel_pos_clip, self.rel_pos_clip)
+
+            final_mat = distance_mat_clipped + self.rel_pos_clip
+
+            rel_pos_embeddings = self.rel_pos_embeddings[final_mat]  # [T, T', F']
+        else:
+            rel_pos_embeddings = self._sinusoidal_pe(
+                torch.arange(time_dim_size - 1, -time_dim_size, -1, device=input_tensor.device, dtype=torch.float32),
+                self.embed_dim_per_head,
+            ).expand(
+                time_dim_size, 2 * time_dim_size - 1, self.embed_dim_per_head
+            )  # [T, T+T'-1, F']
+
+        # dropout relative positional embeddings
+        rel_pos_embeddings = self.pos_emb_dropout(rel_pos_embeddings)
+
+        q_with_bias_u = q + self.pos_bias_u if self.with_pos_bias else q  # [B, T, #heads, F']
+        q_with_bias_v = q + self.pos_bias_v if self.with_pos_bias else q
+
+        # attention matrix a and c
+        attn_ac = torch.einsum("bihf, bjhf -> bhij", q_with_bias_u, k)  # [B, #heads, T, T']
+
+        # attention matrix b and d
+        attn_bd = torch.einsum("bihf, ijf -> bhij", q_with_bias_v, rel_pos_embeddings)
+
+        if not self.learnable_pos_emb:
+            attn_bd = self._rel_shift_bhij(attn_bd, k_len=time_dim_size)
+
+        attn = attn_ac + attn_bd + mask  # [B, #heads, T, T']
+        attn_scaled = attn * (math.sqrt(1.0 / float(self.embed_dim_per_head)))  # [B, #heads, T, T']
+
+        # softmax and dropout
+        attn_output_weights = self.att_weights_dropout(F.softmax(attn_scaled, dim=-1))  # [B, #heads, T, T']
+
+        # sequence of weighted sums over value sequence
+        v = value_seq.view(batch_dim_size, -1, self.num_heads, self.embed_dim_per_head)  # [B, T, H, F']
+        attn_output = (
+            torch.einsum("bhij, bjhf -> bhif", attn_output_weights, v)
+            .transpose(1, 2)
+            .contiguous()
+            .view(batch_dim_size, -1, self.embed_dim)
+        )
+
+        output_tensor = self.out_proj(attn_output)
+
+        if self.dropout_broadcast_axes is None:
+            output_tensor = self.dropout(output_tensor)
+        elif self.dropout_broadcast_axes == "T":
+            output_tensor = self.dropout(output_tensor.transpose(1, 2)).transpose(1, 2)
+        elif self.dropout_broadcast_axes == "B":
+            output_tensor = self.dropout(output_tensor.permute(1, 2, 0)).permute(2, 0, 1)
+        elif self.dropout_broadcast_axes == "BT":
+            batch_dim_size = output_tensor.shape[0]
+            feature_dim_size = output_tensor.shape[-1]
+
+            output_tensor = (
+                self.output(output_tensor.reshape(-1, feature_dim_size).transpose(0, 1))
+                .transpose(0, 1)
+                .reshape(batch_dim_size, -1, feature_dim_size)
+            )
```

Comment:
Again simplify this.
```suggestion
        output_tensor = self.dropout(output_tensor)
```